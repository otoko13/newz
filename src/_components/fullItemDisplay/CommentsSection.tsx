import React from 'react';
import './commentsSection.scss';	
import NewsService, { INewsItem } from '../newsService';
import LoadShimmer from '../loadShimmer/LoadShimmer';
import Comment from './Comment';

export interface ICommentsSectionProps {	
    newsItem?: INewsItem;
    isFirstLevel?: boolean;
}	

const CommentsSection = (props: ICommentsSectionProps) => {
    const [comments, setComments] = React.useState<INewsItem[]>([]);
    const [loading, setLoading] = React.useState(false);

    React.useEffect(() => {
        if (props.newsItem) {
            setLoading(true);
            NewsService.getComments(props.newsItem).then(comments => {
                setLoading(false);
                setComments(comments);
            }).catch(() => {
                setLoading(false);
            });
        }
    }, [props.newsItem]);

    return (
        <div className={`CommentsSection ${props.isFirstLevel ? '' : 'nested'}`}>
            {
                loading && <LoadShimmer />
            }
            {
                (props.isFirstLevel && !loading && comments.length === 0) &&
                <div className='no-comments-message'>No Comments</div>	
            }
            { 
                (!loading && comments.length > 0) &&
                comments.map(comment => <Comment key={comment.id} comment={comment} />)
            }
        </div>	
    );	
}	

export default CommentsSection; 