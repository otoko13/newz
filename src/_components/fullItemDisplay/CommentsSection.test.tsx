import { act, render } from '@testing-library/react';
import * as React from 'react';
import CommentsSection, { ICommentsSectionProps } from './CommentsSection';
import NewsService, { INewsItem } from '../newsService';

jest.mock('../newsService', () => {
    const mockComments: INewsItem[] = [
        { id: 2, descendants: 0, by: 'Maurice', kids: [], time: 1581442689, title: 'Another comment', text: 'I believe', type: 'story' },
        { id: 3, descendants: 0, by: 'Phillip', kids: [], time: 1581442689, title: 'Yet another comment', text: 'I don\'t', type: 'story' },
        { id: 4, descendants: 0, by: 'Kevin', kids: [], time: 1581442689, title: 'One more comment', text: 'Well who the hell are you', type: 'story' },
    ]

    return {
        getComments: jest.fn().mockResolvedValue(mockComments),
    }
});

let props: ICommentsSectionProps;

beforeEach(() => {
    props = {
        newsItem: {
            id: 1,
            descendants: 0,
            by: 'King Julian',
            kids: [2, 3, 4],
            time: 1581442689,
            title: 'A comment',
            text: 'I believe',
            type: 'story',
        }
    }
});

describe('snapshot', () => {
    it('should look sensible', async () => {
        const { container } = render(<CommentsSection {...props} />);
        expect(container.firstChild).toMatchSnapshot();
        await act(async () => {
            await NewsService.getComments({} as INewsItem);
        });
    });
});

describe('loading', () => {
    it('should show a loading shimmer when comments are not yet loaded', async () => {
        const { queryByTitle } = render(<CommentsSection {...props} />);
        expect(queryByTitle('loading')).not.toBeNull();
        await act(async () => {
            await NewsService.getComments({} as INewsItem);
            expect(queryByTitle('loading')).toBeNull();
        });
    });
});

describe('no comments', () => {
    it('should not show a no comments message when there are comments', async () => {
        const { queryByText } = render(<CommentsSection {...props} />);
        await act(async () => {
            await NewsService.getComments({} as INewsItem);
            expect(queryByText('No Comments')).toBeNull();
        });
    });

    it('should show a no comments message when there are no comments and this is the first level of comments', async () => {
        NewsService.getComments = jest.fn().mockResolvedValue([]);
        const { queryByText } = render(<CommentsSection {...props} isFirstLevel={true} />);
        await act(async () => {
            await NewsService.getComments({} as INewsItem);
            expect(queryByText('No Comments')).not.toBeNull();
        });
    });

    it('should not show a no comments message when there are no comments and this is not the first level of comments', async () => {
        NewsService.getComments = jest.fn().mockResolvedValue([]);
        const { queryByText } = render(<CommentsSection {...props} isFirstLevel={false} />);
        await act(async () => {
            await NewsService.getComments({} as INewsItem);
            expect(queryByText('No Comments')).toBeNull();
        });
    });
});
