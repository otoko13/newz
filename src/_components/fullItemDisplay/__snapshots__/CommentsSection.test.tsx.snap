// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should look sensible 1`] = `
<div
  class="CommentsSection"
>
  <div
    class="title"
  >
    Comments:
  </div>
  <div
    class="LoadShimmer"
    style="height: 95px; margin-bottom: 20px;"
    title="loading"
  >
    <div
      class="ms-Shimmer-container top root-38"
    >
      <div
        class="ms-Shimmer-shimmerWrapper shimmerWrapper-39"
        style="width: 100%;"
      >
        <div
          class="ms-Shimmer-shimmerGradient shimmerGradient-40"
        />
        <div
          class="ms-ShimmerElementsGroup-root root-43"
          style="width: auto;"
        >
          <div
            class="ms-ShimmerLine-root root-44"
            style="width: 80%; min-width: auto;"
          >
            <svg
              class="ms-ShimmerLine-topLeftCorner topLeftCorner-45"
              height="2"
              width="2"
            >
              <path
                d="M0 2 A 2 2, 0, 0, 1, 2 0 L 0 0 Z"
              />
            </svg>
            <svg
              class="ms-ShimmerLine-topRightCorner topRightCorner-46"
              height="2"
              width="2"
            >
              <path
                d="M0 0 A 2 2, 0, 0, 1, 2 2 L 2 0 Z"
              />
            </svg>
            <svg
              class="ms-ShimmerLine-bottomRightCorner bottomRightCorner-47"
              height="2"
              width="2"
            >
              <path
                d="M2 0 A 2 2, 0, 0, 1, 0 2 L 2 2 Z"
              />
            </svg>
            <svg
              class="ms-ShimmerLine-bottomLeftCorner bottomLeftCorner-48"
              height="2"
              width="2"
            >
              <path
                d="M2 2 A 2 2, 0, 0, 1, 0 0 L 0 2 Z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ms-Shimmer-container root-38"
    >
      <div
        class="ms-Shimmer-shimmerWrapper shimmerWrapper-39"
        style="width: 30%;"
      >
        <div
          class="ms-Shimmer-shimmerGradient shimmerGradient-40"
        />
        <div
          class="ms-ShimmerElementsGroup-root root-43"
          style="width: auto;"
        >
          <div
            class="ms-ShimmerLine-root root-49"
            style="width: 100%; min-width: auto;"
          >
            <svg
              class="ms-ShimmerLine-topLeftCorner topLeftCorner-45"
              height="2"
              width="2"
            >
              <path
                d="M0 2 A 2 2, 0, 0, 1, 2 0 L 0 0 Z"
              />
            </svg>
            <svg
              class="ms-ShimmerLine-topRightCorner topRightCorner-46"
              height="2"
              width="2"
            >
              <path
                d="M0 0 A 2 2, 0, 0, 1, 2 2 L 2 0 Z"
              />
            </svg>
            <svg
              class="ms-ShimmerLine-bottomRightCorner bottomRightCorner-47"
              height="2"
              width="2"
            >
              <path
                d="M2 0 A 2 2, 0, 0, 1, 0 2 L 2 2 Z"
              />
            </svg>
            <svg
              class="ms-ShimmerLine-bottomLeftCorner bottomLeftCorner-48"
              height="2"
              width="2"
            >
              <path
                d="M2 2 A 2 2, 0, 0, 1, 0 0 L 0 2 Z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
